server/api/include/irods/rs_genquery2.hpp

```diff
@@ -0,0 +1,88 @@
+#ifndef IRODS_RS_GENQUERY2_HPP
+#define IRODS_RS_GENQUERY2_HPP
+
+/// \file
+
+#include "irods/genquery2.h"
+
+struct RsComm;
+
+/// Query the catalog using the GenQuery2 parser.
+///
+/// \warning This API is experimental and may change in the future.
+///
+/// The GenQuery2 parser supports the following:
+/// - Enforces the iRODS permission model
+/// - Logical AND, OR, and NOT
+/// - Grouping via parentheses
+/// - SQL CAST
+/// - SQL GROUP BY
+/// - SQL aggregate functions (e.g. count, sum, avg, etc)
+/// - Per-column sorting via ORDER BY [ASC|DESC]
+/// - SQL FETCH FIRST \a N ROWS ONLY (LIMIT offered as an alias)
+/// - Metadata queries involving different iRODS entities (i.e. data objects, collections, users, and resources)
+/// - Operators: =, !=, <, <=, >, >=, LIKE, BETWEEN, IS [NOT] NULL
+/// - SQL keywords are case-insensitive
+/// - Federation
+/// - Escaping of single quotes
+/// - Bytes encoded as hexadecimal (e.g. \x21)
+///
+/// Limitations:
+/// - Groups are not yet fully supported
+/// - Cannot resolve tickets to data objects and collections using a single query
+/// - Integer values must be treated as strings, except when used for OFFSET, LIMIT, FETCH FIRST \a N ROWS ONLY
+///
+/// When the query does not include the FETCH FIRST \a N ROWS ONLY or LIMIT clause, the API will clamp the
+/// size of the resultset to 256 rows.
+///
+/// The API does not provide any built-in support for pagination due to it being application-dependent.
+/// However, the API does provide the necessary facilities for implementing various forms of pagination.
+///
+/// The column mappings between GenQuery2 and the catalog can be obtained via the API. See the Genquery2Input
+/// structure for details.
+///
+/// \param[in]     _comm   A pointer to a RsComm.
+/// \param[in]     _input  A pointer to a Genquery2Input.
+/// \param[in,out] _output \parblock A pointer that will hold the results of the operation.
```

Comment:
Two reasons:
- Trying to signal to reader that the function expects the caller to pass an object in and the passed object may be modified
- That's how I've written them in the past

I'll check other projects and the doxygen docs for best practice.

I'm open to alternatives as well.

Reply:
I feel like the `in` piece of `in,out` implies that there is anything being supplied by the caller that is used by the API. The only requirement is that the pointer is not `nullptr` so that it can be made to point at another pointer to some memory that is completely manufactured *by* the API as an output for use by the caller.

It feels misleading to me because the API does not use the contents of the pointer at all and in fact immediately dereferences it in order to set the pointed-to-pointer to `nullptr`.

Reply:
I agree. All instances should be updated to use `[out]` instead.

Reply:
Done